conans/client/graph/graph.py

```diff
@@ -335,3 +327,46 @@ def _order_levels(self, direct):
             opened = opened.difference(current_level)
 
         return result
+
+    def private_skip_binaries(self, nodes_subset=None, root=None):
+        """ check which nodes are reachable from the root, mark the non reachable as BINARY_SKIP.
+        Used in the GraphBinaryAnalyzer"""
+        public_nodes = set()
+        root = root if root is not None else self.root
+        nodes = nodes_subset if nodes_subset is not None else self.nodes
+        current = [root]
+        while current:
+            new_current = set()
+            public_nodes.update(current)
+            for n in current:
+                if n.binary in (BINARY_CACHE, BINARY_DOWNLOAD, BINARY_UPDATE, BINARY_SKIP):
```

Comment:
I need some help to understand this `if` 😞  It leaves for the `else` the values of `BINARY_BUILD`, `BINARY_MISSING` and `BINARY_EDITABLE`, shall the _editable_ one be added to the first group?

Also, in the line `to_add = [d.dst for d in n.dependencies if not d.private]`, are all _build-requires_ private? do we want them in this list?

Reply:
Yes, the private deps of BUILD, MISSING and EDITABLE cannot be skipped, we need them to be retrieved.

Build requires are not skipped, if they are in the graph is because they were necessary, otherwise the graph is not expanded to include them.

Reply:
Why the EDITABLE one is different from the CACHE one?

I don't see the difference between build-requires and private here. If privates are not added, why build-requires are? If the binary is available (BINARY_CACHE, DOWNLOAD, UPDATE, SKIP) I understand I don't need anything from the privates but nothing also from the build-requires. Here there is something I'm missing.

Reply:
The CACHE one, the binary is there. Doesn't need to be built. By definition, the EDITABLE is in user space, built by the user, and the user will need all of its dependencies, including the private and the build requires in order to be able to build that binary in edition mode.

Reply:
From my point of view, the EDITABLE binaries should be already there, in the user space, already compiled and available to use. Why it should be different to be in one folder (user space) or in a different one (the cache)?  Building the editable is a different job, a different graph (same graphlock desired) and then the editable is the root of the graph.

Reply:
This is not a change of this PR, this behavior was already there. If the editable has a build-folder, the downstream conan install will generate the ``conanbuildinfo.cmake`` and other files there, so the editable can actually use them to build, with the correct dependencies. Otherwise, the editables won't be able to build, or might even build with different and incompatible dependencies than the ones used to consume it. 
This PR is already a bit risky, but necessary. I prefer not to change other behavior that is not related to the fixes in this one. Using lockfiles for editables is not even tested, not a feature yet, so better leave it as is, and do steps one by one.